### PR TITLE
AJAX Requests to loop over checks to run

### LINKED ISSUE #84: Handle running checks in loop and return output
## Description
Handle running checks one by one and return output.

This issue is dependent on the following:
- #34 
- #86 
- #90 

## Acceptance Criteria
- Update the `run_checks()` method in the `Admin_AJAX` class to handle running checks and returning a response.
    - The `run_checks` method will first verify the request using the nonce passed and If invalid will return an error response.
    - Then the `AJAX_Runner` instance is gathered by calling the `Plugin_Request_Utility::get_runner()`.
    - If the `Plugin_Request_Utility::get_runner()` returns `null` a new instance of the `AJAX_Runner` class should be created.
        - As part of creating the `AJAX_Runner` instance, the `set_check_slugs()` and `set_plugin_slug ()` methods should be called passing in the appropriate values taken from the request. 
    - Then the `AJAX_Runners` `run()` method is called and the `Check_Results` instance returned is stored in a variable.
    - If the `run()` method throws an exception, the error message is captured and returned in an error response to the client. The `admin-plugin-check.js` file will handle the error and display it on the admin page for debugging. Any further check requests will not be sent. 
    - The results are then passed to the protected `format_results()` method in the `Admin_AJAX` class. This method will handle formatting all results into an array of messages, with each message being an object with the following properties.
        - `message` - The error or warning message.
        - `type` - Either `error` or `warning`.
        - `standard` - The standard or check name.
        - `file` - The filename and path
        - `line_column` - The line and column numbers the error/warning occurred.
    - A response is returned containing the following information
        - `success` - `true` or `false` as to whether the check run was successful.
        - `results` - The array of messages returned by `format_results()`. Can be an empty array.
        - `check` - The slug of the check that was run.
        - `error` - `false` if there were no critical errors when running checks, `false` if the check run was successful.
- Update the `plugin-check-admin.js` to send requests to the `admin-ajax.php` endpoint to run checks.
    - The `checks_to_run` returned by the `plugin_check_get_checks_to_run` action will be used to send requests for each individual check.
    - The first check is removed from the array and sent in the request.
    - The AJAX request will send the following:
        - A array containing a single check slug to be run
        - The plugin name to check.
        - A valid nonce.
    - On a successful response:
        - The messages returned will be injected into the admin page using the templates setup in #35 
        - The next check in the `checks_to_run` array  is removed and used in the next request. This process will repeat until the `checks_to_run` is empty and the last check is run.
         - When  all checks have been run and the `checks_to_run` array is empty the runtime environment cleanup is executed by making a request to the `admin-ajax.php` endpoint with the `plugin_check_cleanup_environment` action (implemented in #86)
    - On a error response:
        - A user-friendly error message is displayed in the admin page with the error that occurred.
        - The runtime environment cleanup function will be run by making a request to the `admin-ajax.php` endpoint with the `plugin_check_cleanup_environment` action (implemented in #86)

### PR DESCRIPTION
**Pull request by jjgrainger**
Addresses acceptance criteria in #84 

Closes #84

### CODE REVIEW

**Comment by jjgrainger** [comment]
Set up as a draft so once #104 is merged I can update with the JS formatting here

File: assets/js/plugin-check-admin.js [hunk 1]
```diff
@@ -149,15 +149,24 @@
...
 	 * @since n.e.x.t
 	 */
-	function runChecks( data ) {
+	async function runChecks( data ) {
+		const results = [];
+
+		for (var i = 0; i < data.checks.length; i++) {
+			let result = await runCheck( data.plugin, data.checks[ i ] );
+			results.push( result );
+		}
```
Comment (spacedmonkey) [inline assets/js/plugin-check-admin.js]: As a rule, it is a not good to do a series of promises in a loop like this. Do these NEED to run it order. What if one of the promises fail?

Comment (spacedmonkey) [reply assets/js/plugin-check-admin.js]: I would recommend using `Promise.all` here. 

Try something like this. 

```
const promises = [];
for (var i = 0; i < data.checks.length; i++) {
	promises.push(  runCheck( data.plugin, data.checks[ i ] ) );
}
const results = await Promise.all( promises );
```

File: includes/Admin/Admin_AJAX.php [hunk 2]
```diff
@@ -248,4 +279,107 @@ protected function has_runtime_check( array $checks ) {
...
+			}
+
+			$file_results = $this->flatten_file_results( $file_name, $file_errors, $file_warnings );
+			$all_results  = array_merge( $all_results, $file_results );
+		}
+
+		// If there are any files left with only warnings, print those next.
+		foreach ( $warnings as $file_name => $file_warnings ) {
+			$file_results = $this->flatten_file_results( $file_name, array(), $file_warnings );
+			$all_results  = array_merge( $all_results, $file_results );
```
Comment (spacedmonkey) [inline includes/Admin/Admin_AJAX.php]: It can be expensive to call `array_merge`. Maybe just add `$file_results` to an array and then fatten it at the end.

File: includes/Admin/Admin_AJAX.php [hunk 3]
```diff
@@ -248,4 +279,107 @@ protected function has_runtime_check( array $checks ) {
...
+				}
+				if ( $a['line'] > $b['line'] ) {
+					return 1;
+				}
+				if ( $a['column'] < $b['column'] ) {
+					return -1;
+				}
+				if ( $a['column'] > $b['column'] ) {
+					return 1;
+				}
```
Comment (spacedmonkey) [inline includes/Admin/Admin_AJAX.php]: These seems repeated, how about something like this. 

```
	if ( $a['line'] < $b['line'] || $a['column'] < $b['column'] ) {
					return -1;
	}
```

Comment (felixarntz) [reply includes/Admin/Admin_AJAX.php]: I don't think that would be the same logic, we have to _first_ consider both conditions around the `line` before we look at `column`.

File: includes/Admin/Admin_AJAX.php [hunk 4]
```diff
@@ -208,9 +209,39 @@ public function run_checks() {
...
+
+		// Make sure we are using the correct runner instance.
+		if ( ! ( $runner instanceof AJAX_Runner ) ) {
+			wp_send_json_error(
+				new WP_Error( 'invalid-runner', __( 'AJAX Runner was not initialized correctly.', 'plugin-check' ) ),
+				500
+			);
+		}
+
+		$checks = isset( $_REQUEST['checks'] ) ? array_filter( $_REQUEST['checks'] ) : array();
```
Comment (felixarntz) [inline includes/Admin/Admin_AJAX.php]: Why is that done differently here from above? I would think we should use the same logic everywhere for this parameter.
```suggestion
		$checks = filter_input( INPUT_POST, 'checks', FILTER_DEFAULT, FILTER_FORCE_ARRAY );
		$checks = is_null( $checks ) ? array() : $checks;
```

File: assets/js/plugin-check-admin.js (hunk 1 above)
Comment (felixarntz) [reply assets/js/plugin-check-admin.js]: Hmm, not sure about that one here. Normally I would agree with you, but I think it's a better user experience to show the check results in a consistent order, so circling through the checks makes sense to me in this context.

If we use `Promise.all`, they will be handled in a random order. That would only be a good idea if we wanted to output all results together at the end, but I think it's a good idea to output results one by one as it gives the user feedback already as soon as there are results.

File: assets/js/plugin-check-admin.js [hunk 5]
```diff
@@ -149,15 +149,24 @@
...
+
+		console.log( results );
```
Comment (felixarntz) [inline assets/js/plugin-check-admin.js]: See my comment above, if you're doing this, then I would agree with @spacedmonkey's feedback. However, I would suggest we render the output from each check as soon as we have it. In other words, we should prepare for that here, i.e. `console.log` every individual result right away rather than putting them all in an array and then outputting them.

File: includes/Admin/Admin_AJAX.php [hunk 6]
```diff
@@ -208,9 +209,39 @@ public function run_checks() {
...
+				new WP_Error( 'invalid-request', $error->getMessage() ),
+				400
+			);
+		}
+
 		wp_send_json_success(
 			array(
-				'message' => __( 'Verified!', 'plugin-check' ),
+				'results' => $this->format_results( $results ),
+				'checks'  => $checks,
```
Comment (felixarntz) [inline includes/Admin/Admin_AJAX.php]: Do we need to return `checks` from here? Why is that needed?

File: includes/Admin/Admin_AJAX.php [hunk 7]
```diff
@@ -248,4 +279,107 @@ protected function has_runtime_check( array $checks ) {
...
+				unset( $warnings[ $file_name ] );
+			}
+
+			$file_results = $this->flatten_file_results( $file_name, $file_errors, $file_warnings );
+			$all_results  = array_merge( $all_results, $file_results );
+		}
+
+		// If there are any files left with only warnings, print those next.
+		foreach ( $warnings as $file_name => $file_warnings ) {
+			$file_results = $this->flatten_file_results( $file_name, array(), $file_warnings );
```
Comment (felixarntz) [inline includes/Admin/Admin_AJAX.php]: Why do we need to "flatten" the results? I don't see why that's needed. If it is helpful for the output, we should probably do that in the other issue in JS, but I'm not sure why we would do that.

I'm not even sure we need to do anything to the formatting here. Shouldn't this be the responsibility of the JS? The way this data is best formatted depends on the way it is being rendered, and that should be decoupled from the AJAX endpoint. I think we could probably avoid all this logic and instead simply return `errors` and `warnings` with the raw data in the AJAX response.

File: includes/Admin/Admin_AJAX.php [hunk 8]
```diff
@@ -208,9 +209,39 @@ public function run_checks() {
...
+			wp_send_json_error(
+				new WP_Error( 'invalid-request', $error->getMessage() ),
+				400
+			);
+		}
+
 		wp_send_json_success(
 			array(
-				'message' => __( 'Verified!', 'plugin-check' ),
+				'results' => $this->format_results( $results ),
```
Comment (felixarntz) [inline includes/Admin/Admin_AJAX.php]: See my other comment: I don't think we need to format to cater for a specific opinionated output format, we can return the raw data here and then handle formatting as needed in JS, according to how we use the template.
```suggestion
				'errors    => $results->get_errors(),
				'warnings' => $results->get_warnings(),
```

**Review by felixarntz — COMMENTED** [review]
Thanks @jjgrainger, mostly looks good. My main feedback is that I don't think we should do all the formatting in the AJAX endpoint as it seems unnecessary here and could happen in JS depending on the template usage.

File: assets/js/plugin-check-admin.js (hunk 1 above)
Comment (jjgrainger) [reply assets/js/plugin-check-admin.js]: Thanks @spacedmonkey @felixarntz 

Another point to raise here is that we would want these to run sequentially. We wouldn't want to have 2 runtime checks running in parallel, where both their preparations are implemented at the same time, as this may cause incorrect results returned by checks.

I have updated this to log the results as they're returned.

File: includes/Admin/Admin_AJAX.php (hunk 2 above)
Comment (jjgrainger) [reply includes/Admin/Admin_AJAX.php]: Thanks @spacedmonkey I agree here, although the updates to handle the formatting on the JS side now has meant this code has been removed.

File: includes/Admin/Admin_AJAX.php (hunk 4 above)
Comment (jjgrainger) [reply includes/Admin/Admin_AJAX.php]: @felixarntz good catch! Not sure why this was reverted but have updated this to match.

File: assets/js/plugin-check-admin.js (hunk 5 above)
Comment (jjgrainger) [reply assets/js/plugin-check-admin.js]: Updated to log the individual results right away.

File: includes/Admin/Admin_AJAX.php (hunk 6 above)
Comment (jjgrainger) [reply includes/Admin/Admin_AJAX.php]: @felixarntz this was part of the AC's but I agree that we should only return whats required in the response and we can always add things as and when we need it. I've removed the `checks` for now.

File: includes/Admin/Admin_AJAX.php (hunk 7 above)
Comment (jjgrainger) [reply includes/Admin/Admin_AJAX.php]: @felixarntz this was part of the AC's to format the results in a particular way. However, I agree that we should decouple the logic here and return the raw `errors` and `warnings`. The formatting methods have been removed.

File: includes/Admin/Admin_AJAX.php (hunk 8 above)
Comment (jjgrainger) [reply includes/Admin/Admin_AJAX.php]: Thanks @felixarntz updated.

**Comment by jjgrainger** [comment]
Thanks @spacedmonkey & @felixarntz

I've made updates based on the feedback and is ready for you to review.

File: package-lock.json [hunk 9]
```diff
@@ -12237,4 +12237,4 @@
       "dev": true
     }
   }
-}
+}
```
Comment (felixarntz) [inline package-lock.json:12240]: This change seems unrelated?

**Review by felixarntz — APPROVED** [review]
@jjgrainger Thank you for the updates, LGTM!

One tiny thing unrelated to the actual PR.

File: assets/js/plugin-check-admin.js [hunk 10]
```diff
@@ -129,19 +132,33 @@
...
 	 * Run Checks.
 	 *
+	 * @since n.e.x.t
+	 *
 	 * @param {Object} data The response data.
+	 */
+	async function runChecks( data ) {
+		for ( let i = 0; i < data.checks.length; i++ ) {
+			const result = await runCheck( data.plugin, data.checks[ i ] );
+			console.log( result );
```
Comment (spacedmonkey) [inline assets/js/plugin-check-admin.js]:
```suggestion
            try {
			   const result = await runCheck( data.plugin, data.checks[ i ] );
			   console.log( result );
			} catch (e) {
              // Ignore for now.
             }   
```

My point about error handling seems to have been missed.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Requested one change and I will approve. We are nearly there.

File: package-lock.json (hunk 9 above)
Comment (jjgrainger) [reply package-lock.json:12240]: @felixarntz this would have come about when resolving merge conflicts in when bringing in the JS linting. Looks like my editor automatically removed the newline at end of file which matches the `.editorconfig` we have.

File: assets/js/plugin-check-admin.js (hunk 10 above)
Comment (jjgrainger) [reply assets/js/plugin-check-admin.js]: Thanks @spacedmonkey this has been updated.

**Review by spacedmonkey — APPROVED** [review]
Looking good to me.
